pkg/scheduler/framework/interface.go

```diff
@@ -370,6 +370,10 @@ type EnqueueExtensions interface {
 	// and leveraged to build event handlers dynamically.
 	// Note: the returned list needs to be static (not depend on configuration parameters);
 	// otherwise it would lead to undefined behavior.
+	//
+	// The returned events could be nil to indicate that no events other than the pod's own update
+	// can make the pod re-schedulable. An example is SchedulingGates plugin.
```

Comment:
This comment will conflict with the future direction being discussed in https://github.com/kubernetes/kubernetes/pull/119607#discussion_r1349233182. To summarize the discussion there, we'll use QueueingHint to decide whether the Pod is moved from unschedQ to activeQ/backoffQ when the pod itself is updated.
So, if the pod's own update can make the pod reschedulable, we should filter out other pods' update in queueinghint.

I'm not saying this change to schedulinggate won't work or have some potential bugs, but almost all changes in this PR need to be reverted in https://github.com/kubernetes/kubernetes/pull/119607.

Reply:
> So, if the pod's own update can make the pod reschedulable, we should filter out other pods' update in queueinghint. (after that PR)

Yes, we can do that, but isn't that less performant, esp. for SchedulingGates? as that means SchedulingGates queueingHint func would be triggered by all pods' update and evaluate everytime.

> This comment conflicts with the future direction being discussed in [#119607 (comment)](https://github.com/kubernetes/kubernetes/pull/119607#discussion_r1349233182).

Are you trying to say that for pod's update, we may **enforce** all plugins to implement fine-grained queueHint functions along with PodUpdate event registration? If so, I would argue that some plugin may prefer generic [isPodUpdated()](https://github.com/kubernetes/kubernetes/blob/0bff705acd8982e34b937116eb2016c9d6e4c4a6/pkg/scheduler/internal/queue/scheduling_queue.go#L615) logic to re-queue pods. This logic has been worked for releases, it doesn't look right to force all plugins to turn to "mandatory PodUpdate event + queueHint".

Reply:
> I'm not sure many plugins need to have this actually 

Actually, this is the motivation of the idea in the [discussion](https://github.com/kubernetes/kubernetes/pull/119607#discussion_r1349233182) I mentioned earlier.
Probably only a few plugins' failure could be solved by Pod itself updates, but we're moving Pods from unschedQ to activeQ/backoffQ when any changes to Pods happen no matter which plugins rejected those Pods in a past scheduling cycle.

Reply:
what if SchedulingGates don't implement queueingHint (iow, it relies on generic `isPodUpdated()` logic)?  is there any side effect?

Reply:
With `PodUpdate` or `PodItselfUpdate` (if we create it), Scheduling Gate doesn't need to implement QHint because QHint would be very similar logic to PreEnqueue.

Reply:
I think I was talking something confusing.

> So, if the pod's own update can make the pod reschedulable, we should filter out other pods' update in queueinghint. 

This is not talking about SchedulingGate particularly, but more generally. (because 
pkg/scheduler/framework/interface.go is not only for SchedulingGate)
For example, TaintToleration - it needs to have PodUpdate with QHint and QHint needs to filter out all events but ones which add/update toleration.
And like TaintToleration, if the update of Pod itself could solve the scheduling failure of the plugin, the plugin has to have QHint to filter out other pods' updates.

But, talking particularly about SchedulingGate, I think we don't need QHint for PodUpdate because QHint would be very similar to PreEnqueue.

Reply:
What we do here is **make it possible** for plugins to ignore all events since will not make the scheduling performance better, like schedulingGates. And I think it's not conflict with https://github.com/kubernetes/kubernetes/pull/119607 for preEnqueue will not trigger any pod been moved into `inFlightPods`, right? This only happens when Pop from the activeQ.

Reply:
It _was_ already possible to ignore all events -- as pointed out in https://github.com/kubernetes/kubernetes/pull/121677/files#r1384548595, there's no change in the scheduler side by this PR.

And, the conflict is the comment change on pkg/scheduler/framework/interface.go. 
>	// The returned events could be nil to indicate that no events other than the pod's own update
	// can make the pod re-schedulable. An example is SchedulingGates plugin.

The returned events _couldn't_ be nil if the pod's own update can make the pod re-schedulable. SchedulingGate needs to have PodUpdated event again when https://github.com/kubernetes/kubernetes/pull/119607#discussion_r1349233182 is done.

Reply:
> SchedulingGate needs to have PodUpdated event again when https://github.com/kubernetes/kubernetes/pull/119607#discussion_r1349233182 is done.

I think a Pod failed by schedulingGate plugin will not enter into the `inFlightPods` for this only happens in `priorityQueue.Pop` action, right? So it's not necessary to have `podUpdate` event, have a hintFunc is also ok but less efficient as a second option I talked with Wei offline before.

Besides, how to solve @Huang-Wei 's concern here, this is also what I care about

> This logic has been worked for releases, it doesn't look right to force all plugins to turn to "mandatory PodUpdate event + queueHint".

This is counter-intuitively also seems like a breaking change considering out-of-tree plugins, although I do think https://github.com/kubernetes/kubernetes/pull/119607 is on the right way, but maybe we should think of a better way and not impact endusers.

Reply:
> I think a Pod failed by schedulingGate plugin will not enter into the inFlightPods 

What I'm talking about is completely nothing to do with in-flight Pods.

Reply:
> Besides, how to solve @Huang-Wei 's concern here, this is also what I care about

The concern doesn't have point, if SchedulingGate has PodUpdated event, it works as it is. As I said, SchedulingGate doesn't need to have QHint.

Reply:
> This is counter-intuitively also seems like a breaking change considering out-of-tree plugins, 

But, this is true. And, that's why I started this discussion here.
Before this PR, we hadn't said that the plugin doesn't need to have PodUpdated if they just want to move Pods with unsched Pod itself update. But, this PR clarify the current behavior and that behavior will probably be changed soon.
Anyways, we have to, at least, call out for out-of-tree plugins developers when we do https://github.com/kubernetes/kubernetes/pull/119607#discussion_r1349233182.

Reply:
Can we revert this change for `pkg/scheduler/framework/interface.go`? 
I'm not against this PR overall as it works with the current PR, we can change SchedulingGate again later.
But, I don't want to clarify the current behavior of the scheduling queue that it always moves Pods from unschedQ to active/backoff when unsched Pods are changed. Because this behavior is supposed to be changed very soon.

Reply:
Fine with me revert the comments in interface.go.

Reply:
PR: https://github.com/kubernetes/kubernetes/pull/121948

@kerthcet can you change the release note to NONE?

Reply:
Let me understand first, so we'll process with https://github.com/kubernetes/kubernetes/pull/119607 anyway and shout out the mitigation as registering `podUpdate` for plugins which will make the pod reschedulable by its own update event, right? 

E.g. for schedulingGates, we'll register the `podUpdate` event and **better to** add the hintFunc to filter out unrelated pod update events for performance consideration.

And the main risk here is if a plugin doesn't register the `podUpdate` event, a pod may stay in the unschedulable pods until the background flush as https://github.com/kubernetes/kubernetes/pull/119607#discussion_r1364476601 discussed here.

Reply:
Yes, correct.

Reply:
Or the alternative idea would be:
1. Create a new event `PodItselfUpdate`, which is event for unschedulable Pod itself update.
2. If EventsToRegister doesn't have `PodItselfUpdate`, we add `PodItselfUpdate` to EventsToRegister with no QHint. (so that we don't break all existing plugins.)
3. If `PodItselfUpdate` never makes unsched Pods schedulable, plugins need to have `PodItselfUpdate` with QHint which always returns QueueSkip.

It's safer because it wouldn't break existing plugins, but it may be not intuitive (thus not better) because (3) is something different from other events.

Reply:
We should have a podOwnUpdate event registered by default for each plugin and if a plugin already registered, leave it as what we do today.

Reply:
> We should have a podOwnUpdate event registered by default for each plugin and if a plugin already registered, leave it as what we do today.

Yes. 

And we have to distinguish whether people just don't know about podOwnUpdate or people don't want to requeue by podOwnUpdate. The former, they need a default podOwnUpdate, and the latter, they don't need a default podOwnUpdate.
That's why I mentioned (3) in the above. To distinguish the former and the latter, in the latter, people put podOwnUpdate along with QHint which always returns QueueSkip. (And this is a bit intuitive because, in other cluster events, "event with QHint which always returns QueueSkip" is just the same as "not registering the event in EventsToRegister in the first place".)

Reply:
Hope to hear from others as well because both solutions are not that perfect.

Reply:
Let's move this discussion to https://github.com/kubernetes/kubernetes/pull/119607#discussion_r1349233182. Aldo knows the context too.

Reply:
I'll make a comment there to summarize the discussion here and tag you as well